sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2322,7 +2322,7 @@ class Dataset[T] private[sql](
     }
     val attrs = this.logicalPlan.output
     val colsAfterDrop = attrs.filter { attr =>
-      attr != expression
+      !attr.semanticEquals(expression)
```

Comment:
```
  def semanticEquals(other: Expression): Boolean =
    deterministic && other.deterministic && canonicalized == other.canonicalized
```

What is the reason the comparison should be related to the deterministic when we want to drop it?

Reply:
nvm. The output only contains Attribute